File: chapter07/lists_and_recursion_1_2_3_4.ex
defmodule MyList do
  def mapsum(list, func), do: _mapsum(list, func, 0)

  defp _mapsum([], _, acc), do: acc
  defp _mapsum([head | tail], func, acc),
    do: _mapsum(tail, func, func.(head) + acc)

  def max([head | tail]), do: _max(tail, head)

  defp _max([], acc), do: acc
  defp _max([head | tail], acc) when head > acc, do: _max(tail, head)
  defp _max([_ | tail], acc), do: _max(tail, acc)

  def caesar([], _), do: []
  def caesar([head | tail], n) when head + n > ?z,
    do: [head + n - (?z - ?a + 1) | caesar(tail, n)]
  def caesar([head | tail], n), do: [head + n | caesar(tail, n)]

  def span(from, to) when from > to, do: []
  def span(from, to), do: [from | span(from + 1, to)]
end

File: chapter07/lists_and_recursion_1_2_3_4.clj
(ns lists-and-recursion-1-2-3-4)

(defn- -mapsum [f acc [head & tail :as coll]]
  (if (empty? coll)
    acc
    (recur f (+ (f head) acc) tail)))

(defn mapsum [f coll]
  (-mapsum f 0 coll))

(defn- -max [acc [head & tail :as coll]]
  (cond
    (empty? coll) acc
    (> head acc) (recur head tail)
    :else (recur acc tail)))

(defn max' [[head & tail]] (-max head tail))

(defn caesar [[head & tail :as s] n]
  (cond
    (empty? s) ""

    (> (+ (int head) n) (int \z))
    (str (char (- (+ (int head) n)
                  (inc (- (int \z) (int \a)))))
         (caesar tail n))

    :else (str (char (+ (int head) n))
               (caesar tail n))))

(defn span [from to]
  (if (> from to)
    ()
    (cons from (span (inc from) to))))
